//
//  Interceptor.swift
//  Roomie
//
//  Created by ÍπÄÏäπÏõê on 5/29/25.
//

import Foundation

import Alamofire

final class Interceptor: RequestInterceptor {
    
    static let shared = Interceptor()
    private init() {}
    
    let retryLimit = 3
    
    func adapt(
        _ urlRequest: URLRequest,
        for session: Session,
        completion: @escaping (Result<URLRequest, Error>) -> Void
    ) {
        print("‚û°Ô∏è[Interceptor] - adapt")
        var urlRequest = urlRequest
        
        if let accessToken = TokenManager.shared.fetchAccessToken() {
            if urlRequest.headers.dictionary["Authorization"] == nil {
                urlRequest.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
            }
        }
        
        completion(.success(urlRequest))
    }
    
    func retry(
        _ request: Request,
        for session: Session,
        dueTo error: Error,
        completion: @escaping (
            RetryResult
        ) -> Void
    ) {
        print("üîÑ[Interceptor] - retry start")
        guard let response = request.task?.response as? HTTPURLResponse,response.statusCode == 401 else {
            return completion(.doNotRetryWithError(error))
        }
        
        guard request.retryCount < retryLimit else {
            return completion(.doNotRetryWithError(error))
        }

        Task {
            do {
                let accessToken = try await authReissue()
                print("‚ñ∂Ô∏è[Interceptor] - retry success, token reissued")
                TokenManager.shared.saveTokens(accessToken: accessToken)
                completion(.retry)
                
            } catch let error as TokenError {
                print("‚è∏Ô∏è[Interceptor] - retry failed: \(error.description)")
                switch error {
                case .noRefreshToken, .refreshTokenExpired, .userNotFound:
                    NotificationCenter.default.post(name: Notification.shouldLogout, object: nil)
                    await Toast.show(message: "ÏÑ∏ÏÖòÏù¥ ÎßåÎ£åÎêòÏóàÏñ¥Ïöî. Îã§Ïãú Î°úÍ∑∏Ïù∏Ìï¥Ï£ºÏÑ∏Ïöî")
                    break
                case .reissueFailed:
                    await Toast.show(message: "ÏÑúÎ≤Ñ Ïò§Î•òÏûÖÎãàÎã§. Îã§Ïãú ÏãúÎèÑÌï¥Ï£ºÏÑ∏Ïöî")
                    break
                case .unknownError:
                    await Toast.show(message: "ÏöîÏ≤≠Ïóê Ïã§Ìå®ÌñàÏäµÎãàÎã§. Ïû†Ïãú ÌõÑ Îã§Ïãú ÏãúÎèÑÌï¥Ï£ºÏÑ∏Ïöî")
                    break
                }
                
                completion(.doNotRetryWithError(error))
            } catch {
                await Toast.show(message: "ÏòàÍ∏∞Ïπò Î™ªÌïú Ïò§Î•òÍ∞Ä Î∞úÏÉùÌñàÏñ¥Ïöî")
                completion(.doNotRetryWithError(error))
            }
        }
    }
}

extension Interceptor {
    enum TokenError: Error {
        case noRefreshToken
        case reissueFailed
        case refreshTokenExpired
        case userNotFound
        case unknownError(error: Error)
        
        var description: String {
            switch self {
            case .noRefreshToken:
                return "Ï†ÄÏû•Îêú refreshTokenÏù¥ ÏóÜÏäµÎãàÎã§."
            case .reissueFailed:
                return "ÌÜ†ÌÅ∞ Ïû¨Î∞úÍ∏â ÏöîÏ≤≠Ïóê Ïã§Ìå®ÌñàÏäµÎãàÎã§."
            case .refreshTokenExpired:
                return "refresthTokenÏù¥ ÎßåÎ£åÎêòÏóàÏäµÎãàÎã§."
            case .userNotFound:
                return "Ìï¥Îãπ Ïú†Ï†ÄÎ•º Ï∞æÏùÑ Ïàò ÏóÜÏäµÎãàÎã§."
            case .unknownError:
                return "Ïïå Ïàò ÏóÜÎäî ÏóêÎü¨Í∞Ä Î∞úÏÉùÌñàÏäµÎãàÎã§."
            }
        }
    }
    
    private func authReissue() async throws -> String {
        do {
            guard let refreshToken = TokenManager.shared.fetchRefreshToken() else {
                throw TokenError.noRefreshToken
            }
            
            guard let response = try await AuthService().authReissue(refreshToken: refreshToken) else {
                throw TokenError.reissueFailed
            }
            
            guard let data = response.data else {
                switch response.code {
                case 40102:
                    throw TokenError.refreshTokenExpired
                case 40403:
                    throw TokenError.userNotFound
                default:
                    throw TokenError.reissueFailed
                }
            }
            
            return data.accessToken
            
        } catch let error as TokenError {
            throw error
        } catch {
            throw TokenError.unknownError(error: error)
        }
    }
}
